src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1188 @@
+        /// </param>
+        /// <exception cref="ArgumentNullException"><paramref name="collection"/> is null.</exception>
+        public OrderedDictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey>? comparer)
+        {
+            ArgumentNullException.ThrowIfNull(collection);
+
+            int capacity = collection is ICollection<KeyValuePair<TKey, TValue>> c ? c.Count : 0;
+            _dictionary = new(capacity, comparer);
+            _list = new(capacity);
+
+            AddRange(collection);
+        }
+
+        /// <summary>Gets the <see cref="IEqualityComparer{TKey}"/> that is used to determine equality of keys for the dictionary.</summary>
+        public IEqualityComparer<TKey> Comparer => _dictionary.Comparer;
+
+        /// <summary>Gets the number of key/value pairs contained in the <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        public int Count => _dictionary.Count;
+
+        /// <inheritdoc/>
+        bool ICollection<KeyValuePair<TKey, TValue>>.IsReadOnly => false;
+
+        /// <inheritdoc/>
+        bool IDictionary.IsReadOnly => false;
+
+        /// <inheritdoc/>
+        bool IList.IsReadOnly => false;
+
+        /// <inheritdoc/>
+        bool IDictionary.IsFixedSize => false;
+
+        /// <inheritdoc/>
+        bool IList.IsFixedSize => false;
+
+        /// <summary>Gets a collection containing the keys in the <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        public KeyCollection Keys => _keys ??= new(this);
+
+        /// <inheritdoc/>
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Keys;
+
+        /// <inheritdoc/>
+        ICollection<TKey> IDictionary<TKey, TValue>.Keys => Keys;
+
+        /// <inheritdoc/>
+        ICollection IDictionary.Keys => Keys;
+
+        /// <summary>Gets a collection containing the values in the <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        public ValueCollection Values => _values ??= new(this);
+
+        /// <inheritdoc/>
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+
+        /// <inheritdoc/>
+        ICollection<TValue> IDictionary<TKey, TValue>.Values => Values;
+
+        /// <inheritdoc/>
+        ICollection IDictionary.Values => Values;
+
+        /// <inheritdoc/>
+        bool ICollection.IsSynchronized => false;
+
+        /// <inheritdoc/>
+        object ICollection.SyncRoot => ((ICollection)_dictionary).SyncRoot;
+
+        /// <inheritdoc/>
+        object? IList.this[int index]
+        {
+            get => GetAt(index);
+            set
+            {
+                ArgumentNullException.ThrowIfNull(value);
+
+                if (value is not KeyValuePair<TKey, TValue> tpair)
+                {
+                    throw new ArgumentException(SR.Format(SR.Arg_WrongType, value, typeof(KeyValuePair<TKey, TValue>)), nameof(value));
```

Comment:
Do we care that this throw hasn't been factored into a stub method?

Reply:
I don't think so, until perf tests prove otherwise. Other collections in System.Collections have open-coded throws rather than using throw helpers for everything.